Sources/SwiftGRPCNIOInteroperabilityTests/TestServiceProvider_NIO.swift

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import SwiftGRPCNIO
+import NIO
+
+/// A service prodiver for the gRPC interoperaability test suite.
+///
+/// See: https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md#server
+public class TestServiceProvider_NIO: Grpc_Testing_TestServiceProvider_NIO {
+  public init() { }
+
+  /// Features that this server implements.
+  public static var implementedFeatures: Set<ServerFeature> {
+    return [
+      .emptyCall,
+      .unaryCall,
+      .streamingOutputCall,
+      .streamingInputCall,
+      .fullDuplexCall,
+      .echoStatus
+    ]
+  }
+
+  /// Server implements `emptyCall` which immediately returns the empty message.
+  public func emptyCall(
+    request: Grpc_Testing_Empty,
+    context: StatusOnlyCallContext
+  ) -> EventLoopFuture<Grpc_Testing_Empty> {
+    return context.eventLoop.makeSucceededFuture(Grpc_Testing_Empty())
+  }
+
+  /// Server implements `unaryCall` which immediately returns a `SimpleResponse` with a payload
+  /// body of size `SimpleRequest.responseSize` bytes and type as appropriate for the
+  /// `SimpleRequest.responseType`.
+  ///
+  /// If the server does not support the `responseType`, then it should fail the RPC with
+  /// `INVALID_ARGUMENT`.
+  public func unaryCall(
+    request: Grpc_Testing_SimpleRequest,
+    context: StatusOnlyCallContext
+  ) -> EventLoopFuture<Grpc_Testing_SimpleResponse> {
+    let promise = context.eventLoop.makePromise(of: Grpc_Testing_SimpleResponse.self)
+    if request.shouldEchoStatus {
+      let code = StatusCode(rawValue: numericCast(request.responseStatus.code)) ?? .unknown
+      promise.fail(GRPCStatus(code: code, message: request.responseStatus.message))
```

Comment:
How about `return context.eventLoop.makeFailedFuture` instead? (Also for all the other cases.)